qiskit/synthesis/clifford/clifford_decompose_ag.py

```diff
@@ -53,7 +58,13 @@ def synth_clifford_ag(clifford: Clifford) -> QuantumCircuit:
         return _decompose_clifford_1q(clifford.tableau)
 
     # Compose a circuit which we will convert to an instruction
-    circuit = QuantumCircuit(clifford.num_qubits, name=str(clifford))
+    qreg = QuantumRegister(clifford.num_qubits)
+    if use_dag:
+        circuit = DAGCircuit()
+        circuit.name = str(clifford)
+        circuit.add_qreg(qreg)
+    else:
+        circuit = QuantumCircuit(qreg, name=str(clifford))
```

Comment:
Generally this looks fine, yet I do find that this introduces code duplication and that the new code a bit harder to read and to maintain. 

How much time do we save (here and in other synthesis functions) by directly constructing a `DAGCircuit` rather than constructing a `QuantumCircuit` and converting it to `DAGCircuit` using `dag_to_circuit`? 

For instance, If in practice this time is negligible, then maybe we don't really need the `DAGCircuit` version of the code and we can just let `HighLevelSynthesis` convert it to `DAGCircuit` (when needed).

And if this time is not negligible, then what exactly causes the slowdown? For instance, would it be possible to create some kind of "list of gates" instead of (the presumably more expensive) `QuantumCircuit`/`DAGCircuit`, and only construct the right object from this list at the very end? 

Alternatively, maybe we can split `synth_clifford_ag` into two functions like `_synth_clifford_ag_circuit`  and `_synth_clifford_ag_dag` that produce `QuantumCircuit` and `DAGCircuit` respectively? (Unfortunately this would introduce even more code duplication but imho would make each individual function more readable).

Or maybe we can have some generic functions like `_append_cx` or `_append_h`, etc. that under the hood use the correct method of adding a gate to either a `DAGCircuit` or a `QuantumCircuit` (and I think we might already have something of this kind).

Also this somewhat relates to the question/task of whether we eventually want to port these algorithms to Rust.

What do you think?

Reply:
I see your point @alexanderivrii, and I agree that the current code is less readable. I am working on adding a HLS plugin benchmark to our ASV suite to be able to test if we are actually getting relevant performance improvements or not, but I haven't had time to finish it yet. I think that the best course of action would be to finish the benchmark and make a decision based off the results. I also considered the option to have a full dag-based function as an alternative, but it's a lot of functions to migrate and I wanted to get some feedback on the current state first.